### PR TITLE
8255070: Shenandoah: Use single thread for concurrent CLD liveness test	

### PR DESCRIPTION
**Pull request by zhengyu123**
Since the concurrent liveness test does not touch any oops, there is no point to use multi-thread for the task, they just duplicate the work.

Also, multi-thread version requires to acquire ClassLoaderGraph_lock early, which is deadlock prone.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255070](https://bugs.openjdk.java.net/browse/JDK-8255070): Shenandoah: Use single thread for concurrent CLD liveness test

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/777/head:pull/777`
`$ git checkout pull/777`

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -1941,7 +1941,7 @@ class ShenandoahConcurrentWeakRootsEvacUpdateTask : public AbstractGangTask {
   ShenandoahVMWeakRoots<true /*concurrent*/> _vm_roots;
 
   // Roots related to concurrent class unloading
-  ShenandoahClassLoaderDataRoots<true /* concurrent */, false /* single thread*/>
+  ShenandoahClassLoaderDataRoots<true /* concurrent */, true /* single thread*/>
```
Comment (AzeemJiva) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1944]: Should the first input be false?

Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1944]: No. It runs concurrently, so will need to acquire ClassLoaderGraph_lock.

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp [hunk 2]
```diff
@@ -81,25 +82,33 @@ ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::ShenandoahClassLoad
...
+        MutexLocker locker(ClassLoaderDataGraph_lock, Mutex::_no_safepoint_check_flag);
+        ClassLoaderDataGraph::always_strong_cld_do(clds);
+        _semaphore.claim_all();
+      }
+    } else {
+      assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
+      assert(Thread::current()->is_VM_thread(), "Can only be done by VM thread");
+      ClassLoaderDataGraph::always_strong_cld_do(clds);
+    }
   } else if (_semaphore.try_acquire()) {
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp:112]: It looks to me that we do:
if (_semaphore.try_acquire()) {

in both branches now. Can this be reshaped to do try_acquire() and claim_all() outermost?

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp [hunk 3]
```diff
@@ -110,9 +119,18 @@ void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::always_strong_
...
+        MutexLocker locker(ClassLoaderDataGraph_lock, Mutex::_no_safepoint_check_flag);
+        ClassLoaderDataGraph::cld_do(clds);
+        _semaphore.claim_all();
+      }
+    } else {
+      assert(SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
+      assert(Thread::current()->is_VM_thread(), "Can only be done by VM thread");
+      ClassLoaderDataGraph::cld_do(clds);
+    }
   } else if (_semaphore.try_acquire()) {
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp:134]: Same as above. Also, the only difference to above is cld_do() vs always_strong_cld_do(), with *lots* of identical boilerplate. Can this be refactored into a generic method that takes the final target method as argument?

**Review by rkennke — COMMENTED** [review]
Also, I'd change the commit message to begin uppercase after Shenandoah: "Use..."

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp (hunk 3 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp]: Refactored and retested. Thanks for the suggestion.

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp [hunk 4]
```diff
@@ -81,43 +81,46 @@ ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::ShenandoahClassLoad
...
 
+template <bool CONCURRENT, bool SINGLE_THREADED>
+void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::cld_do_impl(CldDo f, CLDClosure* clds, uint worker_id) {
+  if (CONCURRENT) {
+    if (_semaphore.try_acquire()) {
+      ShenandoahWorkerTimingsTracker timer(_phase, ShenandoahPhaseTimings::CLDGRoots, worker_id);
+      if (SINGLE_THREADED) ClassLoaderDataGraph_lock->lock();
+      f(clds);
+      if (SINGLE_THREADED) ClassLoaderDataGraph_lock->unlock();
+      _semaphore.claim_all();
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp:108]: Would it be nicer to write this as follows?
if (SINGLE_THREADED) {
  MutexLocker ml(ClassLoaderDataGraph_lock);
  f(clds);
} else {
  f(clds);
}

**Review by rkennke — APPROVED** [review]
Looks good to me! Thank you!

**Pull request closed without merging**